### PR TITLE
Handle line-end comments

### PR DESCRIPTION
**Pull request by boblail**
Allows `psych-comments` to accurately around-trip comments that fall the same line as values — such as

```yaml
fruits:
- apples
- bananas
- tomatoes # requires some explanation
```

### CODE REVIEW

**Comment by boblail** [comment]
@qnighy, do you mind taking a look? (🙏 Thank you!)

**Comment by boblail** [comment]
@qnighy bump

**Comment by boblail** [comment]
@qnighy bump

**Comment by qnighy** [comment]
Thank you for the PR!

Two comments:

- `end_comment` should be something clearly distinguishable from `trailing_comments` by its name e.g. `line_end_comment`.
- It should accept multiple comments to support the following example in the future:
  ```yaml
  foo: bar # line end comment 1
           # line end comment 2
  # trailing comment 1
  # trailing comment 2
  ```

**Comment by boblail** [comment]
@qnighy, Thanks for taking a look!

> `end_comment` should be something clearly distinguishable from `trailing_comments` by its name e.g. `line_end_comment`.

Good call! This is clearer. I renamed that and pushed up a new commit.

> It should accept multiple comments to support the following example in the future:
> ```yaml
> foo: bar # line end comment 1
>          # line end comment 2
> # trailing comment 1
> # trailing comment 2
> ```

Are you suggesting that, in this PR, the property be renamed `line_end_comments`? (with the `s`) and should return an array?

I didn't want to make an API-breaking change with this pull request; but would it be even better to parse line-end comments into the first element of the array of `trailing_comments`?

 - In your example above, `trailing_comments` for the Scalar `bar` would be `["# line end comment 1", "# line end comment 2"]` and the `trailing_comments` for the Mapping (or the Document) would be `["# trailing comment 1", "# trailing comment 2"]`
 - For every node without a line-end comment _or_ a trailing comment, `trailing_comments` would be `[]`
 - For every node without a line-end comment but _with_ a trailing comment, `trailing_comments[0]` would be `nil` or `""`

**Comment by qnighy** [comment]
> Are you suggesting that, in this PR, the property be renamed `line_end_comments`? (with the `s`) and should return an array?

Yes.

> I didn't want to make an API-breaking change with this pull request

I'm sorry for that but that is how contribution works. And as you are changing `end` to `line_end` here, a breaking change is technically inevitable here.

> would it be even better to parse line-end comments into the first element of the array of `trailing_comments`?

Yeah, line-end comments complicates things and multi-line ones do even much. However, I want to make less API-breaking changes with this library going forward. Thus I want to incorporate sufficient future-proofness into the API before getting it merged.

**Comment by boblail** [comment]
> > Are you suggesting that, in this PR, the property be renamed line_end_comments? (with the s) and should return an array?
> 
> Yes.

OK! I pushed up that change.

File: lib/psych/comments/emitter.rb
Comment (qnighy) [inline lib/psych/comments/emitter.rb]: Could you throw away whitespace details before calling `emit_comment`? The library is meant to preserve comments, but not the spacing details.

**Review by qnighy — CHANGES_REQUESTED** [review]

File: lib/psych/comments/emitter.rb
Comment (boblail) [reply lib/psych/comments/emitter.rb]: We certainly could!

But I don't think we should.

It would wreck `psych-comments`' ability to round-trip YAML like this (with justified line-end comments):
```yaml
fruit:
- fig      # ripens in late spring
- apricot  # ripens in the summer
- apple    # ripens in the fall
```

It wouldn't be able to faithfully render the YAML it parses. I'd think that fidelity would be paramount, wouldn't you?

Comment (qnighy) [reply lib/psych/comments/emitter.rb]: The library is meant to preserve comments, but not the spacing details.

If you want justification, you may implement a justification algorithm and I'm willing to accept it if its quality is good enough. Major design decisions you will need to consider implementing it are:

- How is the width calculated? What happens if there is ambiwidth or long grapheme cluster?
- What constitutes a single "run" of comments that should be aligned at the same position? Should it be limited to the same level of leaves? What happens if there are lines with comments and those without?

Comment (boblail) [reply lib/psych/comments/emitter.rb]: Hey Masaki!

Oh, I definitely wouldn't want to introduce an algorithm for justification — I wouldn't expect this library to have an opinion about whitespace! I wouldn't want it to _introduce_ justification where it didn't exist before. (I didn't see it like Rubocop or prettier.js or anything that would _reformat_ YAML.)

My understanding is that psych-comments is trying to allow you to manipulate YAML more losslessly (from the README: "This gem allows you to manipulate YAML, preserving comment information.")

My use-case is manipulating thousands of YAML files written by as many engineers. They (my internal customers) have different conventions for spacing and commenting these configuration files.

What are the cons to preserving whatever whitespace exists between Scalar nodes and line-end comments?
What would be the advantage of discarding it?

Comment (qnighy) [reply lib/psych/comments/emitter.rb]: > I didn't see it like Rubocop or prettier.js or anything that would reformat YAML.

It **does** behave like a reformatter to a certain degree -- consider the following example:

```yaml
-           foo: \"\\x62\\x61\\x72\"
```

When you `Psych::Comments.emit_yaml(Psych::Comments.parse_stream(source))` it, this reformats to:

```yaml
- foo: "bar"
```

The behavior is meant to imitate how Psych (or libyaml) behaves.

> My understanding is that psych-comments is trying to allow you to manipulate YAML more losslessly

No, this is as it reads -- it preserves comment information.

> What are the cons to preserving whatever whitespace exists between Scalar nodes and line-end comments?

Missing a clear boundary on the library's scope sometimes puts a heavy burden on the maintainer, and _I_ cannot take the risk in this case. As you say, many YAML authors have many opinions on how they write the value, the prominent example of which is

```yaml
foo:
- bar
- baz
```

vs.

```yaml
foo:
  - bar
  - baz
```

or whatever details. As I'm not prepared to accept all these requirements, I'm going to decide following principles:

- The library is meant to preserve comments, but not the spacing details.
- Or, to put it differently: this is Psych (libyaml) augmented with comment node. Other syntactic details are thrown away.

If you are confident you can handle (i.e. accept or reject) all those sorts of additional demands and keep the library clean enough to maintain, I won't stop you from forking this library (nor have the authority to do so, of course). Though the gem named `psych-comments` will remain simple enough to handle just comments.

Comment (boblail) [reply lib/psych/comments/emitter.rb]: I pushed up that change.

I'll have to go a different direction for my use-case; but if that wraps up this PR, we can merge it. If not, we can close it.

File: spec/emitter/example4-in.yml
Comment (qnighy) [inline spec/emitter/example4-in.yml:4]: Could you add another test case rather than modifying this one?

File: spec/emitter/example3-in.yml
Comment (qnighy) [inline spec/emitter/example3-in.yml:2]: Could you add another test case rather than modifying this one?

File: lib/psych/comments/node_ext.rb
Comment (qnighy) [inline lib/psych/comments/node_ext.rb:19]: This change has two problems:

- The effect of this overriding is somewhat unpredictable when nodes with comments are passed to a tool unaware of comments.
- The behavior is inconsistent with the other parts of this library, as only line_end_comments are honored.

Could you drop this change for now? I'm willing to discuss this change after the PR is merged. Though in my opinion, adding newly named methods seems better than overriding existing ones.

**Review by qnighy — COMMENTED** [review]
Sorry for being late. The patch is almost perfect except the three things:

**Comment by boblail** [comment]
@samrjenkins, I'm going to decline this PR, but feel free to open a new one with your changes!

**Comment by qnighy** [comment]
Thank you for the PR anyway!

**Pull request closed without merging**
